compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/FocusSwitcher.desktop.kt

```diff
@@ -19,86 +19,110 @@ package androidx.compose.ui.awt
-    ) {
-        private val requester = FocusRequester()
+private class FocusTracker(
+    properties: FocusProperties.() -> Unit,
+    private val onFocus: (FocusDirection) -> Unit,
+) {
+    private val requester = FocusRequester()
+    private var isRequestingFocus = false
+    private var lastEnteredDirection = FocusDirection.Enter
 
-        private var isRequestingFocus = false
-        private var isHandlingFocus = false
+    fun requestFocusWithoutEvent() {
+        try {
+            isRequestingFocus = true
+            requester.requestFocus()
+        } finally {
+            isRequestingFocus = false
+        }
+    }
 
-        fun requestFocusWithoutEvent() {
-            try {
-                isRequestingFocus = true
-                requester.requestFocus()
-            } finally {
-                isRequestingFocus = false
+    private val childModifier = Modifier
+        .focusProperties(properties)
+        .focusRequester(requester)
+        .onFocusEvent {
+            if (!isRequestingFocus && it.isFocused) {
+                // `parentModifier.onEnter` is always called before `childModifier.onFocusEvent`,
+                // so we always have the actual value
+                onFocus(lastEnteredDirection)
             }
         }
+        .focusTarget()
 
-        val modifier = Modifier
-            .focusRequester(requester)
-            .onFocusEvent {
-                if (!isRequestingFocus && !isHandlingFocus && it.isFocused) {
-                    try {
-                        isHandlingFocus = true
-                        onNonRecursiveFocused()
-                    } finally {
-                        isHandlingFocus = false
-                    }
-                }
+    private val parentModifier = Modifier
+        .focusProperties {
+            canFocus = false
+            onEnter = {
+                lastEnteredDirection = requestedFocusDirection
             }
-            .focusTarget()
-    }
+        }
+        .focusTarget()
+
+    // 2 modifiers:
+    // - parent to intercept onEnter, but without focusing on it
+    // - child to handle the focus
+    //
+    // The logic parent/child is got from [Modifier.focusInteropModifier].
+    //
+    // The difference with Android is that the [childModifier] is requesting the SwingPanel focus,
+    // not the [parentModifier].
+    //
+    // The reason for that is the case when SwingPanel is at the beginning of the window.
+    // In this case we call `onRequestFocusForOwner` when we focus on the first node:
+    //
+    //     parentModifier.onEnter -> onRequestFocusForOwner -> childModifier.onFocusEvent
+    //
+    // If we would also place focusing on a SwingPanel component between
+    // onEnter an onRequestFocusForOwner, onRequestFocusForOwner will override the focus.
+    //
+    // The issue may exist on Android too.
+    val modifier = Modifier
+        .then(parentModifier)
+        .then(childModifier)
```

Comment:
Is the split to two modifiers just for ease-of-reading/cleanliness? It seems the only place the two are used is right here, so there's no technical need to have two.